Adding GitHub 101 and ST Instructions (#1)
fhdsl/ITN_Workshops_2024: github-activity-instructions.Rmd

<input type="checkbox"> Click the blue `Clone` button.  

### Window Juggling

Note the three buttons on the main GitHub Desktop repository page that connect us to related apps.

We will be "juggling" some windows when doing this activity:

- `Code editor` -- we'll actually edit files with this
- `Files` -- we'll find where our files are on our computer with this
- `Online GitHub` -- we will send our pull request and suggested changes to online GitHub.com with this.

### Filing a pull request

Pull requests are great for suggesting changes in a way that allow us to preview and discuss the changes before merging to the `main` version.

We will practice filing a pull request by filling in the the README. READMEs are great for getting folks up to speed on a repository or code base.

### Create a new branch

But to create a pull request, we first need to create a branch to work from that is separate from the `main` copy.

<input type="checkbox"> Click the `Current Branch` box > `New branch`.  
<input type="checkbox"> Call this new branch `readme-edit`.  

### Create an AI generated README

> <input type="checkbox"> Click the `Show in Finder` button.  

Reviewer: On a windows machine the button says "Show in Explorer"

Author: ```suggestion
<input type="checkbox"> Click the `Show in Finder` or `Show in Explorer` button depending on your operating system.  
```

<input type="checkbox"> With any text editor open up both the `01-heatmap.Rmd` file and the `README.md` file  
<input type="checkbox"> Copy all the text from `01-heatmap.Rmd` (Cmd + A)  
<input type="checkbox"> Go to phind https://www.phind.com/search  
<input type="checkbox"> Ask find to create a README from the code and paste in the code from `01-heatmap.Rmd`. You may want to do some variations on the prompt you use to ask phind to do this to get good results.   
<input type="checkbox"> Copy all the text from the desired Phind output and paste it to `README.md`. Save the file.  

### Adding changes to a branch

> <input type="checkbox"> Go back to GitHub Desktop. You should see `1 changed file` in the upper right corner.  

Reviewer: Like in the image on slide 84

<input type="checkbox"> It should open up a page on GitHub.com Fill out the `Add a title` and the `Add a description` boxes to appropriately describe the work you did on the branch we are making the pull request from.
<input type="checkbox"> Click `Create pull request`.

## Activity 2: Navigating PR components

### 2a: Leaving comments and suggestions

<input type="checkbox"> Return to your pull request from Activity 1.   
<input type="checkbox"> Click on the `Files Changed` tab.   
<input type="checkbox"> Click one of the plus signs to the left to leave a comment.   
<input type="checkbox"> With one of the open comments click the little symbol that has a plus and minus in a page to leave a suggestion.  
<input type="checkbox"> Click `Add a single comment`.  

Note that you or a collaborator can now readily incorporate your suggestion by clicking `Commit suggestion` or they could comment back to further discuss.

<input type="checkbox"> Click `Commit suggestion` and then `Commit changes`.
<input type="checkbox"> Return to the `Conversation` tab and scroll to the bottom. You should see an additional commit now!

### 2b: Reading GitHub Actions logs

> On the bottom of your PR you will notice a box at the bottom that will probably either say `Some checks haven't completed yet` or `All checks passed`.

Reviewer: Mine says "All checks have failed", but I think that's because there's a single check (Render R Example / render-notebooks (pull request))

This is an automatic check run by GitHub Actions which is one of many systems used for Continuous Integration/Continuous Deployment.

<input type="checkbox"> If needed, click `Show all checks`.  
<input type="checkbox"> Click `Details` at the end of the line that says `Render R Example / render-notebooks (pull_request)`.
<input type="checkbox"> Click around in here to investigate what GitHub Actions logs look like.

For more learning about using [GitHub Actions for science see our course](https://hutchdatascience.org/GitHub_Automation_for_Scientists/).

### 2c: Updating your local repository

If you followed the steps in the `Leaving comments and suggestions` section of this activity you will now have at least one commit that is on your online version of your repository but the changes done online won't automatically get to your local repository. To do this we need to `Pull` the latest changes to our computer.

<input type="checkbox"> Return to GitHub Desktop
<input type="checkbox"> Click the `Fetch origin` button.  
<input type="checkbox"> You should be prompted to `Pull origin` click this button.

> <input type="checkbox"> Now your changes should be synced. Check the file you've changed to make sure.

Author: ```suggestion
<input type="checkbox"> Now your changes should be synced. Open up the file you've changed with any default text editor to look for the change you made to make sure.
```

### 2d: Merging your PR

In a real life setting the life of a Pull request could be long or short. What's important is that the discussions that need to be had about the changes are had before the pull request is merged.

Pull requests and GitHub have their own kind of etiquette. Make sure you talk to your team or collaborators to create and discuss what kind of guidelines and workflows you'd like to adhere to so everyone can do their best work!

> <input type="checkbox"> Return to GitHub Pull Request Online

Reviewer: ```suggestion
<input type="checkbox"> Return to GitHub Pull Request Online.  
```

> <input type="checkbox"> Return to GitHub Pull Request Online

Reviewer: When I went back to online GitHub I was in the actions tab (because of where we ended in the last activity). Might need to add explanations on how to get back to the online PR?

Author: Thanks for identifying this problem! We resolved it here: https://github.com/fhdsl/reproducibility-sandbox/commit/969b51555886a6c5f7950fb0fda9d272da092935

> <input type="checkbox"> Click the `Merge` button.

Reviewer: Do we really want to tell them to merge even if the checks have failed? Mine failed because I didn't have a token and it wouldn't render the notebook.

Reviewer: There's also a second "Confirm merge" or "Cancel" button
